### ISSUE TITLE
Allow to plan tests with todo flag

### ISSUE THREAD

**Issue by mateusjbarbosa** [wellwelwel/poku#563]
Hi, everyone!

First, congratulations on the initiative! I’m using Poku in a new project and I’m loving it.

One thing I used in Jest and believe I can add here is also the possibility of planning tests before developing them, using the `todo` flag or method, as the example below.

- Jest format [Jest API reference](https://jestjs.io/pt-BR/docs/api#testtodoname)
```js
test.todo('ensure that is possible sum two numbers')
```

- Propused Poku format
```js
import { describe, test, it } from 'poku'

describe('Using todo', () => {
  test.todo('ensure that is possible sum two numbers')
  // or
  it.todo('ensure that is possible sum two numbers')
})
```

If I may, I can try to develop and open a PR. I would love to be part of the project!

**Comment by wellwelwel** [comment]
Thanks, @mateusjbarbosa 🤝

Can you show a visual reference of what the terminal output would look like?

**Comment by wellwelwel** [comment]
> I'm wondering about the current complexity of `test` / `it` modules:
> https://github.com/wellwelwel/poku/blob/8dae9f3a4113e55885ec9ee261113a7e49185104/src/modules/helpers/test.ts#L8-L14


---

What do you think about this approach?

```ts
import { describe, it, todo } from 'poku'

describe('Using todo', () => {
  todo('ensure that is possible sum two numbers');

  it('real test', () => true)
})

todo('ensure that is possible sum two numbers')
```

?

**Comment by mateusjbarbosa** [comment]
> Thanks, @mateusjbarbosa 🤝
> 
> Can you show a visual reference of what the terminal output would look like?
> 
> For example, the `skip` visual:
> 
> <img alt="Screenshot 2024-07-20 at 07 15 51" width="628" src="https://private-user-images.githubusercontent.com/46850407/350684675-84407519-a719-4b41-b09e-1e9efaa2dc72.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0NzE2NzMsIm5iZiI6MTcyMTQ3MTM3MywicGF0aCI6Ii80Njg1MDQwNy8zNTA2ODQ2NzUtODQ0MDc1MTktYTcxOS00YjQxLWIwOWUtMWU5ZWZhYTJkYzcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIwVDEwMjkzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE0ZTU5YTI1MDY3NzBhMWE1YzcxZjhlNTlhOGE4YjNkMjY2MWI3YjRhZDM3Zjk3MGQ0ZjBmZTFmZDMzZGUzN2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.KlWwNPEVGW608anY8KY60sgyzhn96VjiLxGZwwRLMh8"> <img alt="Screenshot 2024-07-20 at 07 16 40" width="628" src="https://private-user-images.githubusercontent.com/46850407/350684719-05f7594a-7205-4de4-8b93-14363e645c21.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE0NzE2NzMsIm5iZiI6MTcyMTQ3MTM3MywicGF0aCI6Ii80Njg1MDQwNy8zNTA2ODQ3MTktMDVmNzU5NGEtNzIwNS00ZGU0LThiOTMtMTQzNjNlNjQ1YzIxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIwVDEwMjkzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhNmNlMDQyNWI3ZDBjNThmMjJkODA2OGMyOTNmZjk2NTM1MDdiMTgxNzUxMTY0MTU2ZDc1ZDgyYzZlMmIzNDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8HH95Jz_p1_f7Tc6ptX_sAdLqoz6KqF2pnf4ZT73ZIA">
> > As I've never used this method before, in my mind it would sound similar.

Of course, @wellwelwel !

I think it could be something like this:

```zsh

Directory: <dir>

<file>.test.ts
    ✔︎ passed test description
    ✘ failed test description
    ❍ planned test description (yellow or orange)
✔ 1/3 <cmd> › <time>

Start at › <start>
Duration › <duration>
Test Files › <files count>

PASS › 1 PLANNED › 1 (same color) FAIL › 1
```

Of course we can think of a better icon for the planned test, it was just an example.

**Comment by mateusjbarbosa** [comment]
> > _I'm wondering about the current complexity of `test` / `it` modules_
> > https://github.com/wellwelwel/poku/blob/8dae9f3a4113e55885ec9ee261113a7e49185104/src/modules/helpers/test.ts#L8-L14
> 
> What do you think about this approach?
> 
> ```ts
> import { describe, it, todo } from 'poku'
> 
> describe('Using todo', () => {
>   todo('ensure that is possible sum two numbers');
> 
>   it('real test', () => true)
> })
> 
> todo('ensure that is possible sum two numbers')
> ```

It's perfect!

**Comment by mrazauskas** [comment]
To be honest, I just wanted to open an issue asking for `.fail`, `.only`, `.skip`, `.todo` modifiers. Not so important how that would be designed. What is important:

- it must be possible to set any of the modifiers for any `describe()`, `test()` or `it()`
- also a single assertion can be skipped or focused (that’s optional)

Motivation is very simple. If tests are well organised, while debugging / developing certain behaviour user might want to skip or focus the whole file, the whole `describe()` or `test()` and even a single `assert()`. Why not? ;D

**Comment by wellwelwel** [comment]
> (note that test() is not required in Poku). Why not? ;D

In theory, even assertions aren't necessary, it's enough for the process to end in success.
My idea in creating **Poku** was to have a really modular tester that I could build my own style of testing.

Please feel free to request features, report bugs and even discuss ideias for _PRs_ (I'd love that) 🙋🏻‍♂️

**Comment by wellwelwel** [comment]
Hey @mateusjbarbosa, it was solved in #633 and will be released in the next version 🙋🏻‍♂️

Similar to the `.skip` method from other testers, `.todo` also accepts callbacks which will be ignored for future elaborations or prototype implementations.

---

Now that all the `bug` labels have been fixed, I'm happy to be able closing the first `feature request` 🎉

**Comment by mateusjbarbosa** [comment]
Once again, thank you for your attention and congratulations on the work! 🚀

**Comment by wellwelwel** [comment]
> Once again, thank you for your attention and congratulations on the work! 🚀

Thanks for your support, @mateusjbarbosa 💙

> Already released in `v2.3.0` 🙋🏻‍♂️
